File: Go/Decirational/decimal_integer_test.go
package decirational

import "testing"

func mustD(t *testing.T, s string) DecimalInteger {
	t.Helper()
	v, err := ParseDecimalInteger(s)
	if err != nil {
		t.Fatalf("ParseDecimalInteger(%q): %v", s, err)
	}
	return v
}

func expectPanic(t *testing.T, desc string, fn func()) {
	t.Helper()
	defer func() {
		if recover() == nil {
			t.Errorf("%s: expected a panic", desc)
		}
	}()
	fn()
}

// Construction / String
func TestDecimalIntegerConstruction(t *testing.T) {
	cases := []struct{ input, want string }{
		{"123", "123"}, {"-123", "-123"}, {"+123", "123"}, {"000123", "123"},
		{"0", "0"}, {"-0", "0"}, {"123456789012345678901234567890", "123456789012345678901234567890"},
	}
	for _, c := range cases {
		if got := mustD(t, c.input).String(); got != c.want {
			t.Errorf("parse(%q) = %q, want %q", c.input, got, c.want)
		}
	}
	if NewDecimalIntegerFromInt64(0).String() != "0" {
		t.Error("int64 constructor zero")
	}
	if NewDecimalIntegerFromInt64(42).String() != "42" {
		t.Error("int64 constructor positive")
	}
	if NewDecimalIntegerFromInt64(-42).String() != "-42" {
		t.Error("int64 constructor negative")
	}
	if !DecimalZero.Equals(mustD(t, "0")) {
		t.Error("DecimalZero equals parsed zero")
	}
	if !DecimalOne.Equals(mustD(t, "1")) {
		t.Error("DecimalOne equals parsed one")
	}
	if seven, err := GetDecimalDigit(7); err != nil || !seven.Equals(mustD(t, "7")) {
		t.Error("GetDecimalDigit(7) matches parsed 7")
	}

	for _, bad := range []string{"", "-", "12a", "1.5"} {
		if _, err := ParseDecimalInteger(bad); err == nil {
			t.Errorf("expected error parsing %q", bad)
		}
	}
	if _, err := GetDecimalDigit(10); err == nil {
		t.Error("GetDecimalDigit(10) should be rejected")
	}
	if _, err := NewDecimalInteger(nil, false); err == nil {
		t.Error("NewDecimalInteger with empty digits should be rejected")
	}
	if _, err := NewDecimalInteger([]byte{1, 10, 2}, false); err == nil {
		t.Error("NewDecimalInteger with an out-of-range digit should be rejected")
	}
}

func TestDecimalIntegerPredicates(t *testing.T) {
	if !mustD(t, "0").IsZero() || mustD(t, "1").IsZero() {
		t.Error("IsZero")
	}
	if !mustD(t, "1").IsOne() || mustD(t, "-1").IsOne() || mustD(t, "2").IsOne() {
		t.Error("IsOne")
	}
	if !mustD(t, "1").IsUnitAbs() || !mustD(t, "-1").IsUnitAbs() || mustD(t, "2").IsUnitAbs() {
		t.Error("IsUnitAbs")
	}
	if !mustD(t, "5").IsPositive() || mustD(t, "0").IsPositive() || mustD(t, "-5").IsPositive() {
		t.Error("IsPositive")
	}
	if !mustD(t, "-5").IsNegative() || mustD(t, "0").IsNegative() || mustD(t, "5").IsNegative() {
		t.Error("IsNegative")
	}
}

func TestDecimalIntegerNegateAbs(t *testing.T) {
	if got := mustD(t, "5").Negate().String(); got != "-5" {
		t.Errorf("negate positive: %s", got)
	}
	if got := mustD(t, "-5").Negate().String(); got != "5" {
		t.Errorf("negate negative: %s", got)
	}
	if got := mustD(t, "0").Negate().String(); got != "0" {
		t.Errorf("negate zero stays zero: %s", got)
	}
	if got := mustD(t, "-5").Abs().String(); got != "5" {
		t.Errorf("abs of negative: %s", got)
	}
}

func TestDecimalIntegerCompare(t *testing.T) {
	cases := []struct {
		a, b string
		want int // -1, 0, or 1 (sign only compared)
	}{
		{"5", "3", 1}, {"3", "5", -1}, {"5", "5", 0},
		{"-5", "3", -1}, {"-5", "-3", -1}, {"-3", "-5", 1},
		{"100", "99", 1}, {"-100", "-99", -1},
	}
	for _, c := range cases {
		got := mustD(t, c.a).Compare(mustD(t, c.b))
		if sign(got) != c.want {
			t.Errorf("compare(%s,%s) = %d, want sign %d", c.a, c.b, got, c.want)
		}
	}
}

func sign(n int) int {
	if n > 0 {
		return 1
	} else if n < 0 {
		return -1
	}
	return 0
}

func TestDecimalIntegerEquals(t *testing.T) {
	if !mustD(t, "007").Equals(NewDecimalIntegerFromInt64(7)) {
		t.Error("equal values from different literals are equal")
	}
	if mustD(t, "5").Equals(mustD(t, "-5")) {
		t.Error("5 should not equal -5")
	}
}

func TestDecimalIntegerPlusMinus(t *testing.T) {
	cases := []struct{ a, b, want string }{
		{"123", "456", "579"}, {"-5", "3", "-2"}, {"5", "-3", "2"}, {"-5", "-3", "-8"},
		{"0", "7", "7"}, {"7", "0", "7"}, {"5", "-5", "0"},
	}
	for _, c := range cases {
		if got := mustD(t, c.a).Plus(mustD(t, c.b)).String(); got != c.want {
			t.Errorf("%s+%s = %s, want %s", c.a, c.b, got, c.want)
		}
	}
	minusCases := []struct{ a, b, want string }{
		{"5", "3", "2"}, {"3", "5", "-2"}, {"-5", "-3", "-2"}, {"5", "-3", "8"}, {"0", "7", "-7"}, {"7", "0", "7"},
	}
	for _, c := range minusCases {
		if got := mustD(t, c.a).Minus(mustD(t, c.b)).String(); got != c.want {
			t.Errorf("%s-%s = %s, want %s", c.a, c.b, got, c.want)
		}
	}
}

func TestDecimalIntegerMultiply(t *testing.T) {
	cases := []struct{ a, b, want string }{
		{"12", "12", "144"}, {"-3", "4", "-12"}, {"-3", "-4", "12"}, {"0", "999", "0"},
		{"1", "999", "999"}, {"-1", "999", "-999"}, {"999", "-1", "-999"},
	}
	for _, c := range cases {
		if got := mustD(t, c.a).Multiply(mustD(t, c.b)).String(); got != c.want {
			t.Errorf("%s*%s = %s, want %s", c.a, c.b, got, c.want)
		}
	}
}

func TestDecimalIntegerMultiplyDivideBase(t *testing.T) {
	if got := mustD(t, "12").MultiplyBase(2).String(); got != "1200" {
		t.Errorf("multiply_base by 2: %s", got)
	}
	if got := mustD(t, "12").MultiplyBase(0).String(); got != "12" {
		t.Errorf("multiply_base by 0 is identity: %s", got)
	}
	if got := mustD(t, "-12").MultiplyBase(2).String(); got != "-1200" {
		t.Errorf("multiply_base preserves sign: %s", got)
	}
	if got := mustD(t, "1234").DivideByBase(2).String(); got != "12" {
		t.Errorf("divide_by_base by 2: %s", got)
	}
	if got := mustD(t, "1234").DivideByBase(0).String(); got != "1234" {
		t.Errorf("divide_by_base by 0 is identity: %s", got)
	}
	if got := mustD(t, "1234").DivideByBase(10).String(); got != "0" {
		t.Errorf("divide_by_base beyond length is zero: %s", got)
	}
	expectPanic(t, "multiply_base rejects negative times", func() { mustD(t, "5").MultiplyBase(-1) })
	expectPanic(t, "divide_by_base rejects negative times", func() { mustD(t, "5").DivideByBase(-1) })
}

func TestDecimalIntegerDivideModulo(t *testing.T) {
	cases := []struct{ a, b, want string }{
		{"144", "12", "12"}, {"-144", "12", "-12"}, {"7", "2", "3"}, {"-7", "2", "-3"},
	}
	for _, c := range cases {
		if got := mustD(t, c.a).DivideBy(mustD(t, c.b)).String(); got != c.want {
			t.Errorf("%s/%s = %s, want %s", c.a, c.b, got, c.want)
		}
	}
	moduloCases := []struct{ a, b, want string }{{"7", "2", "1"}, {"-7", "2", "-1"}}
	for _, c := range moduloCases {
		if got := mustD(t, c.a).Modulo(mustD(t, c.b)).String(); got != c.want {
			t.Errorf("%s%%%s = %s, want %s", c.a, c.b, got, c.want)
		}
	}
	expectPanic(t, "divide by zero panics", func() { mustD(t, "5").DivideBy(DecimalZero) })
	expectPanic(t, "modulo by zero panics", func() { mustD(t, "5").Modulo(DecimalZero) })
}

func verifyDecimalDivisionIdentity(t *testing.T, aStr, bStr string) {
	t.Helper()
	a, b := mustD(t, aStr), mustD(t, bStr)
	qr := a.DivideByAndModulo(b)
	reconstructed := qr[0].Multiply(b).Plus(qr[1])
	if !reconstructed.Equals(a) {
		t.Errorf("(%s)/(%s): q*b+r = %s, want %s", aStr, bStr, reconstructed, a)
	}
	if !qr[0].Equals(a.DivideBy(b)) {
		t.Errorf("(%s)/(%s): quotient mismatch with DivideBy", aStr, bStr)
	}
	if !qr[1].Equals(a.Modulo(b)) {
		t.Errorf("(%s)/(%s): remainder mismatch with Modulo", aStr, bStr)
	}
}

func TestDecimalIntegerDivisionIdentity(t *testing.T) {
	pairs := [][2]string{{"17", "5"}, {"-17", "5"}, {"17", "-5"}, {"-17", "-5"}, {"100", "7"}, {"0", "5"}, {"999999999999999999999", "37"}, {"1", "1"}}
	for _, p := range pairs {
		verifyDecimalDivisionIdentity(t, p[0], p[1])
	}
}

func TestDecimalIntegerGcdLcm(t *testing.T) {
	cases := []struct{ a, b, want string }{
		{"48", "18", "6"}, {"17", "5", "1"}, {"100", "75", "25"}, {"0", "5", "5"}, {"5", "0", "5"}, {"-12", "18", "6"},
	}
	for _, c := range cases {
		if got := mustD(t, c.a).Gcd(mustD(t, c.b)).String(); got != c.want {
			t.Errorf("gcd(%s,%s) = %s, want %s", c.a, c.b, got, c.want)
		}
	}
	if got := mustD(t, "21").Lcm(mustD(t, "6")).String(); got != "42" {
		t.Errorf("lcm(21,6) = %s", got)
	}
	if got := mustD(t, "0").Lcm(mustD(t, "5")).String(); got != "0" {
		t.Errorf("lcm(0,x) = %s", got)
	}
}

func TestDecimalIntegerPow(t *testing.T) {
	cases := []struct {
		base string
		exp  int32
		want string
	}{
		{"2", 10, "1024"}, {"5", 0, "1"}, {"0", 0, "1"}, {"5", 1, "5"}, {"-2", 3, "-8"}, {"-2", 2, "4"},
	}
	for _, c := range cases {
		if got := mustD(t, c.base).Pow(c.exp).String(); got != c.want {
			t.Errorf("%s^%d = %s, want %s", c.base, c.exp, got, c.want)
		}
	}
	expectPanic(t, "negative exponent rejected", func() { mustD(t, "2").Pow(-1) })
}

File: Java/Decirational/test/DecimalIntegerTest.java
package decirational;

public final class DecimalIntegerTest {
    private static void verifyDivisionIdentity(final TestFramework t, final String a_str, final String b_str) {
        final DecimalInteger a = new DecimalInteger(a_str);
        final DecimalInteger b = new DecimalInteger(b_str);
        final DecimalInteger[] qr = a.divideByAndModulo(b);
        final DecimalInteger reconstructed = qr[0].multiply(b).plus(qr[1]);
        t.checkEquals(a, reconstructed, "divideByAndModulo(" + a_str + "," + b_str + ") satisfies q*b+r==a");
        t.checkEquals(qr[0], a.divideBy(b), "divideByAndModulo(" + a_str + "," + b_str + ") quotient matches divideBy");
        t.checkEquals(qr[1], a.modulo(b), "divideByAndModulo(" + a_str + "," + b_str + ") remainder matches modulo");
    }

    public static TestFramework run() {
        final TestFramework t = new TestFramework("DecimalInteger");

        // Construction / toString
        t.checkEquals("123", new DecimalInteger("123").toString(), "parse plain digits");
        t.checkEquals("-123", new DecimalInteger("-123").toString(), "parse negative digits");
        t.checkEquals("123", new DecimalInteger("+123").toString(), "parse explicit plus sign");
        t.checkEquals("123", new DecimalInteger("000123").toString(), "strip leading zeros");
        t.checkEquals("0", new DecimalInteger("0").toString(), "zero stays zero");
        t.checkEquals("0", new DecimalInteger("-0").toString(), "negative zero normalizes to zero");
        t.checkEquals("0", new DecimalInteger(0).toString(), "int constructor zero");
        t.checkEquals("42", new DecimalInteger(42).toString(), "int constructor positive");
        t.checkEquals("-42", new DecimalInteger(-42).toString(), "int constructor negative");
        t.checkEquals("123456789012345678901234567890", new DecimalInteger("123456789012345678901234567890").toString(), "large number round-trips");
        t.checkEquals(DecimalInteger.ZERO, new DecimalInteger("0"), "ZERO constant equals parsed zero");
        t.checkEquals(DecimalInteger.ONE, new DecimalInteger("1"), "ONE constant equals parsed one");
        t.checkEquals(DecimalInteger.getDigit(7), new DecimalInteger("7"), "getDigit(7) matches parsed 7");

        t.checkThrows(NumberFormatException.class, () -> new DecimalInteger((String) null), "null string rejected");
        t.checkThrows(NumberFormatException.class, () -> new DecimalInteger(""), "empty string rejected");
        t.checkThrows(NumberFormatException.class, () -> new DecimalInteger("-"), "bare sign rejected");
        t.checkThrows(NumberFormatException.class, () -> new DecimalInteger("12a"), "non-digit character rejected");
        t.checkThrows(NumberFormatException.class, () -> new DecimalInteger("1.5"), "decimal point rejected");
        t.checkThrows(IllegalArgumentException.class, () -> DecimalInteger.getDigit(10), "getDigit out of range rejected");

        // Predicates
        t.check(new DecimalInteger("0").isZero(), "0 is zero");
        t.check(!new DecimalInteger("1").isZero(), "1 is not zero");
        t.check(new DecimalInteger("1").isOne(), "1 is one");
        t.check(!new DecimalInteger("-1").isOne(), "-1 is not one");
        t.check(!new DecimalInteger("2").isOne(), "2 is not one");
        t.check(new DecimalInteger("1").isUnitAbs(), "1 has unit abs");
        t.check(new DecimalInteger("-1").isUnitAbs(), "-1 has unit abs");
        t.check(!new DecimalInteger("2").isUnitAbs(), "2 does not have unit abs");
        t.check(new DecimalInteger("5").isPositive(), "5 is positive");
        t.check(!new DecimalInteger("0").isPositive(), "0 is not positive");
        t.check(!new DecimalInteger("-5").isPositive(), "-5 is not positive");
        t.check(new DecimalInteger("-5").isNegative(), "-5 is negative");
        t.check(!new DecimalInteger("0").isNegative(), "0 is not negative");
        t.check(!new DecimalInteger("5").isNegative(), "5 is not negative");

        // negate / abs
        t.checkEquals("-5", new DecimalInteger("5").negate().toString(), "negate positive");
        t.checkEquals("5", new DecimalInteger("-5").negate().toString(), "negate negative");
        t.checkEquals("0", new DecimalInteger("0").negate().toString(), "negate zero stays zero");
        t.checkEquals("5", new DecimalInteger("-5").abs().toString(), "abs of negative");
        t.checkEquals("5", new DecimalInteger("5").abs().toString(), "abs of positive is unchanged");

        // compareTo
        t.check(new DecimalInteger("5").compareTo(new DecimalInteger("3")) > 0, "5 > 3");
        t.check(new DecimalInteger("3").compareTo(new DecimalInteger("5")) < 0, "3 < 5");
        t.check(0 == new DecimalInteger("5").compareTo(new DecimalInteger("5")), "5 == 5");
        t.check(new DecimalInteger("-5").compareTo(new DecimalInteger("3")) < 0, "-5 < 3");
        t.check(new DecimalInteger("-5").compareTo(new DecimalInteger("-3")) < 0, "-5 < -3");
        t.check(new DecimalInteger("-3").compareTo(new DecimalInteger("-5")) > 0, "-3 > -5");
        t.check(new DecimalInteger("100").compareTo(new DecimalInteger("99")) > 0, "longer positive beats shorter");
        t.check(new DecimalInteger("-100").compareTo(new DecimalInteger("-99")) < 0, "longer negative is smaller");

        // equals / hashCode
        t.checkEquals(new DecimalInteger("007"), new DecimalInteger(7), "equal values from different literals are equal");
        t.check(new DecimalInteger("007").hashCode() == new DecimalInteger(7).hashCode(), "equal values have equal hashCodes");
        t.check(!new DecimalInteger("5").equals(new DecimalInteger("-5")), "5 is not equal to -5");
        t.check(!new DecimalInteger("5").equals(null), "not equal to null");
        t.check(!new DecimalInteger("5").equals("5"), "not equal to different type");
        t.check(new DecimalInteger("5").equals(new DecimalInteger("5")), "equals is reflexive-consistent");

        // plus
        t.checkEquals("579", new DecimalInteger("123").plus(new DecimalInteger("456")).toString(), "123+456");
        t.checkEquals("-2", new DecimalInteger("-5").plus(new DecimalInteger("3")).toString(), "-5+3");
        t.checkEquals("2", new DecimalInteger("5").plus(new DecimalInteger("-3")).toString(), "5+(-3)");
        t.checkEquals("-8", new DecimalInteger("-5").plus(new DecimalInteger("-3")).toString(), "-5+(-3)");
        t.checkEquals("7", new DecimalInteger("0").plus(new DecimalInteger("7")).toString(), "0+7 shortcut");
        t.checkEquals("7", new DecimalInteger("7").plus(new DecimalInteger("0")).toString(), "7+0 shortcut");
        t.checkEquals("0", new DecimalInteger("5").plus(new DecimalInteger("-5")).toString(), "5+(-5)=0");

        // minus
        t.checkEquals("2", new DecimalInteger("5").minus(new DecimalInteger("3")).toString(), "5-3");
        t.checkEquals("-2", new DecimalInteger("3").minus(new DecimalInteger("5")).toString(), "3-5");
        t.checkEquals("-2", new DecimalInteger("-5").minus(new DecimalInteger("-3")).toString(), "-5-(-3)");
        t.checkEquals("8", new DecimalInteger("5").minus(new DecimalInteger("-3")).toString(), "5-(-3)");
        t.checkEquals("-7", new DecimalInteger("0").minus(new DecimalInteger("7")).toString(), "0-7");
        t.checkEquals("7", new DecimalInteger("7").minus(new DecimalInteger("0")).toString(), "7-0");

        // multiply
        t.checkEquals("144", new DecimalInteger("12").multiply(new DecimalInteger("12")).toString(), "12*12");
        t.checkEquals("-12", new DecimalInteger("-3").multiply(new DecimalInteger("4")).toString(), "-3*4");
        t.checkEquals("12", new DecimalInteger("-3").multiply(new DecimalInteger("-4")).toString(), "-3*-4");
        t.checkEquals("0", new DecimalInteger("0").multiply(new DecimalInteger("999")).toString(), "0*999");
        t.checkEquals("999", new DecimalInteger("1").multiply(new DecimalInteger("999")).toString(), "1*999 unit shortcut");
        t.checkEquals("-999", new DecimalInteger("-1").multiply(new DecimalInteger("999")).toString(), "-1*999 unit shortcut");
        t.checkEquals("-999", new DecimalInteger("999").multiply(new DecimalInteger("-1")).toString(), "999*-1 unit shortcut");

        // multiplyBase / divideByBase
        t.checkEquals("1200", new DecimalInteger("12").multiplyBase(2).toString(), "multiplyBase by 2");
        t.checkEquals("12", new DecimalInteger("12").multiplyBase(0).toString(), "multiplyBase by 0 is identity");
        t.checkEquals("-1200", new DecimalInteger("-12").multiplyBase(2).toString(), "multiplyBase preserves sign");
        t.checkEquals("12", new DecimalInteger("1234").divideByBase(2).toString(), "divideByBase by 2");
        t.checkEquals("1234", new DecimalInteger("1234").divideByBase(0).toString(), "divideByBase by 0 is identity");
        t.checkEquals("0", new DecimalInteger("1234").divideByBase(10).toString(), "divideByBase beyond length is zero");
        t.checkThrows(IllegalArgumentException.class, () -> new DecimalInteger("5").multiplyBase(-1), "multiplyBase rejects negative times");
        t.checkThrows(IllegalArgumentException.class, () -> new DecimalInteger("5").divideByBase(-1), "divideByBase rejects negative times");

        // divideBy / modulo
        t.checkEquals("12", new DecimalInteger("144").divideBy(new DecimalInteger("12")).toString(), "144/12");
        t.checkEquals("-12", new DecimalInteger("-144").divideBy(new DecimalInteger("12")).toString(), "-144/12");
        t.checkEquals("3", new DecimalInteger("7").divideBy(new DecimalInteger("2")).toString(), "7/2 truncates");
        t.checkEquals("-3", new DecimalInteger("-7").divideBy(new DecimalInteger("2")).toString(), "-7/2 truncates toward zero");
        t.checkEquals("1", new DecimalInteger("7").modulo(new DecimalInteger("2")).toString(), "7%2");
        t.checkEquals("-1", new DecimalInteger("-7").modulo(new DecimalInteger("2")).toString(), "-7%2 keeps dividend sign");
        t.checkThrows(ArithmeticException.class, () -> new DecimalInteger("5").divideBy(new DecimalInteger("0")), "divide by zero throws");
        t.checkThrows(ArithmeticException.class, () -> new DecimalInteger("5").modulo(new DecimalInteger("0")), "modulo by zero throws");

        verifyDivisionIdentity(t, "17", "5");
        verifyDivisionIdentity(t, "-17", "5");
        verifyDivisionIdentity(t, "17", "-5");
        verifyDivisionIdentity(t, "-17", "-5");
        verifyDivisionIdentity(t, "100", "7");
        verifyDivisionIdentity(t, "0", "5");
        verifyDivisionIdentity(t, "999999999999999999999", "37");
        verifyDivisionIdentity(t, "1", "1");

        // gcd / lcm
        t.checkEquals("6", new DecimalInteger("48").gcd(new DecimalInteger("18")).toString(), "gcd(48,18)");
        t.checkEquals("1", new DecimalInteger("17").gcd(new DecimalInteger("5")).toString(), "gcd of coprimes");
        t.checkEquals("25", new DecimalInteger("100").gcd(new DecimalInteger("75")).toString(), "gcd(100,75)");
        t.checkEquals("5", new DecimalInteger("0").gcd(new DecimalInteger("5")).toString(), "gcd(0,x)=x");
        t.checkEquals("5", new DecimalInteger("5").gcd(new DecimalInteger("0")).toString(), "gcd(x,0)=x");
        t.checkEquals("6", new DecimalInteger("-12").gcd(new DecimalInteger("18")).toString(), "gcd ignores sign");
        t.checkEquals("5", new DecimalInteger("0").gcd(new DecimalInteger("-5")).toString(), "gcd(0,-x)=x");
        t.checkEquals("5", new DecimalInteger("-5").gcd(new DecimalInteger("0")).toString(), "gcd(-x,0)=x");
        t.checkEquals("42", new DecimalInteger("21").lcm(new DecimalInteger("6")).toString(), "lcm(21,6)");
        t.checkEquals("0", new DecimalInteger("0").lcm(new DecimalInteger("5")).toString(), "lcm(0,x)=0");

        // pow
        t.checkEquals("1024", new DecimalInteger("2").pow(10).toString(), "2^10");
        t.checkEquals("1", new DecimalInteger("5").pow(0).toString(), "x^0=1");
        t.checkEquals("1", new DecimalInteger("0").pow(0).toString(), "0^0=1 by convention");
        t.checkEquals("5", new DecimalInteger("5").pow(1).toString(), "x^1=x");
        t.checkEquals("-8", new DecimalInteger("-2").pow(3).toString(), "(-2)^3 is negative");
        t.checkEquals("4", new DecimalInteger("-2").pow(2).toString(), "(-2)^2 is positive");
        t.checkThrows(IllegalArgumentException.class, () -> new DecimalInteger("2").pow(-1), "negative exponent rejected");

        return t;
    }
}
